**Dataset - Reviews.csv**

** Case Problem**

Data Dictionary:

* Id
* ProductId - unique identifier for the product
* UserId - unqiue identifier for the user
* ProfileName
* HelpfulnessNumerator - number of users who found the review helpful
* HelpfulnessDenominator - number of users who indicated whether they found the review helpful
* Score - rating between 1 and 5.
* Time - timestamp for the review
* Summary - brief summary of the review
* Text- review text

In [1]:
import pandas as pd

Read data from Reviews.csv

In [2]:
df = pd.read_csv('Reviews.csv')

In [3]:
df.head(3)

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   

                 Summary                                               Text  
0  Good Quality Dog Food  I have bought several of the Vitality canned d...  
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...  
2  "Delight" says it all  This is a confection that has been around a fe...

Take 'Text' column from dataset and use it to predict the score

In [4]:
from sklearn.feature_extraction.text import CountVectorizer

In [5]:
from sklearn.model_selection import train_test_split

In [6]:
X,y=df['Text'], df['Score']

Split the dataset into train and test at 80:20 ratio

In [7]:
X_train, X_test, y_train, y_test  = train_test_split(X, y, test_size=0.2,random_state=112358)

In [8]:
cvec = CountVectorizer()

In [9]:
X_train_vec = cvec.fit_transform(X_train)

In [10]:
print(X_train_vec)

  (0, 43707)	1
  (0, 69750)	1
  (0, 22750)	1
  (0, 74347)	1
  (0, 81314)	1
  (0, 93222)	1
  (0, 55137)	1
  (0, 54831)	1
  (0, 68364)	1
  (0, 66913)	1
  (0, 103253)	1
  (0, 30718)	1
  (0, 98208)	1
  (0, 44417)	1
  (0, 107501)	1
  (0, 99095)	1
  (0, 91357)	1
  (0, 25058)	1
  (0, 100181)	1
  (0, 7880)	1
  (0, 51472)	1
  (0, 81148)	1
  (0, 54017)	1
  (0, 90247)	1
  (0, 34431)	1
  :	:
  (454761, 56042)	1
  (454761, 97996)	1
  (454761, 98351)	1
  (454762, 67350)	1
  (454762, 85291)	1
  (454762, 41414)	1
  (454762, 44258)	1
  (454762, 18649)	1
  (454762, 65621)	1
  (454762, 105718)	1
  (454762, 61787)	1
  (454762, 88487)	1
  (454762, 106375)	1
  (454762, 106227)	1
  (454762, 66645)	1
  (454762, 98299)	1
  (454762, 68945)	1
  (454762, 48689)	1
  (454762, 22896)	1
  (454762, 78421)	1
  (454762, 56042)	1
  (454762, 97996)	3
  (454762, 107090)	1
  (454762, 98055)	1
  (454762, 57390)	1


In [11]:
X_test_vec = cvec.transform(X_test)

In [12]:
from sklearn.preprocessing import MultiLabelBinarizer

In [13]:
mlb=MultiLabelBinarizer()

In [14]:
y1=y

In [22]:
y1  =y1.reshape([len(y1),1])

In [23]:
label_binarizer=mlb.fit(y1)

In [24]:
y_train= y_train.reshape([len(y_train),1])

In [25]:
y_test= y_test.reshape([len(y_test),1])

In [26]:
y_train_bins = label_binarizer.transform(y_train)

In [27]:
y_test_bins = label_binarizer.transform(y_test)

In [28]:
y_test

array([[1],
       [5],
       [4],
       ...,
       [5],
       [1],
       [1]])

In [29]:
y_test_bins

array([[1, 0, 0, 0, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 0, 1, 0],
       ...,
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0],
       [1, 0, 0, 0, 0]])

### First testing the accuracy with Logistic Regression

In [30]:
from sklearn.multiclass import OneVsRestClassifier

In [31]:
from sklearn.linear_model import LogisticRegression

In [32]:
model = OneVsRestClassifier(LogisticRegression(), n_jobs=-1)

In [33]:
model.fit(X_train_vec.tocoo(), y_train_bins)

OneVsRestClassifier(estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
          n_jobs=-1)

In [34]:
from sklearn.metrics import accuracy_score

In [35]:
acc_score = accuracy_score(y_test_bins,  model.predict(X_test_vec))

In [36]:
print('Model Accuracy: ', acc_score*100, '%')

Model Accuracy:  67.87256686984897 %


### We have obtained accuracy of around 68% with Logistic regression.

### Let's try Naive Bayes theorem and test accuracy

In [37]:
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()

In [46]:
X_train, X_test, y_train, y_test  = train_test_split(X, y, test_size=0.2,random_state=112358)

In [50]:
X_train_vec = cvec.fit_transform(X_train)

In [51]:
X_test_vec = cvec.transform(X_test)

In [52]:
%time nb.fit(X_train_vec, y_train)

CPU times: user 374 ms, sys: 101 ms, total: 475 ms
Wall time: 474 ms


MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [53]:
y_pred_class = nb.predict(X_test_vec)

In [54]:
from sklearn import metrics
print(metrics.accuracy_score(y_test, y_pred_class)*100,'%')

70.21663983956515 %


In [55]:
metrics.confusion_matrix(y_test, y_pred_class)

array([[ 6613,  1036,   747,   423,  1519],
       [ 1439,  1435,  1076,   756,  1262],
       [ 1135,   514,  2769,  1919,  2244],
       [  812,   259,  1115,  6306,  7813],
       [ 2031,   338,  1083,  6340, 62707]])

### We have obtained accuracy of around 70% with Naive Bayes.

### Now we'll keep Naive Bayes and we'll try to have some variations on count vectorizer including n-gram, max_features, stop words etc.
### Let's see if the accuracy improves.

In [56]:
# define a function that accepts a vectorizer and calculates the accuracy
def tokenize_test(vect):
    X_train_dtm = vect.fit_transform(X_train)
    print ('Features: ', X_train_dtm.shape[1])
    X_test_dtm = vect.transform(X_test)
    nb = MultinomialNB()
    nb.fit(X_train_dtm, y_train)
    y_pred_class = nb.predict(X_test_dtm)
    print ('Accuracy: ', metrics.accuracy_score(y_test, y_pred_class)*100,'%')
    print (metrics.confusion_matrix(y_test, y_pred_class))

In [57]:
vect = CountVectorizer(stop_words='english')
tokenize_test(vect)

Features:  108993
Accuracy:  70.19113210368455 %
[[ 6573   822   680   431  1832]
 [ 1416  1286   890   755  1621]
 [ 1104   410  2378  1815  2874]
 [  852   212   809  5612  8820]
 [ 1954   302   926  5365 63952]]


### Accuracy remains 70% with stop_words

In [58]:
vect = CountVectorizer(ngram_range=(1, 2), max_features=100000)
tokenize_test(vect)

Features:  100000
Accuracy:  71.50609986718386 %
[[ 6799  2047   710   243   539]
 [ 1108  3043  1031   407   379]
 [  879  1120  4220  1449   913]
 [  663   629  1684  7868  5461]
 [ 1841   828  1608  8856 59366]]


### Including n-gram and max_features accuracy increased to 71.5%

In [59]:
vect = CountVectorizer(ngram_range=(1, 2), min_df=2)
tokenize_test(vect)

Features:  1277401
Accuracy:  79.01153125577223 %
[[ 7817   271   588   504  1158]
 [ 1169  2058   877   847  1017]
 [  689   124  3950  2052  1766]
 [  402    25   454  8833  6591]
 [  834    59   353  4082 67171]]


### Accuracy increased to 79% by having n-gram and min_df

In [60]:
vect = CountVectorizer(stop_words='english',ngram_range=(1, 2), min_df=2)
tokenize_test(vect)

Features:  1492554
Accuracy:  79.1804100588437 %
[[ 7299   100   324   523  2092]
 [  928  1987   474   831  1748]
 [  569    53  3446  1705  2808]
 [  298    18   205  7886  7898]
 [  603    27   165  2301 69403]]


### We can see slight increase in accuracy (79.18% from 79%) by adding stop_words and keeping n-gram and min_df